src/main/java/uk/gov/hmcts/reform/idam/web/sso/SSOAuthenticationSuccessHandler.java

```diff
@@ -0,0 +1,144 @@
+
+import static org.springframework.http.HttpHeaders.SET_COOKIE;
+import static uk.gov.hmcts.reform.idam.web.helper.ErrorHelper.restException;
+
+@Slf4j
+public class SSOAuthenticationSuccessHandler implements AuthenticationSuccessHandler {
+
+    private final RedirectStrategy redirectStrategy = new DefaultRedirectStrategy();
+
+    private final OAuth2AuthorizedClientRepository repository;
+
+    private final SsoFederationApi federationApi;
+
+    private final OidcApi oidcApi;
+
+    public SSOAuthenticationSuccessHandler(OAuth2AuthorizedClientRepository repository,
+                                           SsoFederationApi federationApi, OidcApi oidcApi) {
+        this.repository = repository;
+        this.federationApi = federationApi;
+        this.oidcApi = oidcApi;
+    }
+
+    @Override
+    public void onAuthenticationSuccess(HttpServletRequest request,
+                                        HttpServletResponse response, Authentication authentication)
+        throws IOException {
+
+        handle(request, response, authentication);
+        clearAuthenticationAttributes(request);
+    }
+
+    protected void handle(
+        HttpServletRequest request,
+        HttpServletResponse response,
+        Authentication authentication
+    ) throws IOException {
+
+        final OAuth2AuthorizedClient client = repository
+            .loadAuthorizedClient(((OAuth2AuthenticationToken) authentication).getAuthorizedClientRegistrationId(),
+                authentication, request);
+
+        final String access_token = client.getAccessToken().getTokenValue();
+        String bearerToken = "Bearer " + access_token;
+
+        updateOrCreateUser(bearerToken);
+
+        String sessionCookie = federationApi.federationAuthenticate(bearerToken)
+            .headers().get(SET_COOKIE).stream()
+            .findAny().orElseThrow(() ->
+                restException(null, HttpStatus.UNAUTHORIZED,  new HttpHeaders(),
+                    HttpStatus.UNAUTHORIZED.getReasonPhrase(), "Unable to authenticate user.")
+            );
+
+        final Map<String, String[]> paramMap = (Map<String, String[]>) request.getSession()
+            .getAttribute("oidcParams");
+
+        if (CollectionUtils.isEmpty(paramMap)) {
+            request.getSession().invalidate();
+            throw restException(null, HttpStatus.FORBIDDEN, new HttpHeaders(),
+                HttpStatus.FORBIDDEN.getReasonPhrase(), "Only federated logins allowed.");
+        }
+
+        final String responseUrl = authoriseUser(Collections.singletonList(sessionCookie), paramMap);
+
+        if (response.isCommitted()) {
+            throw restException(null, HttpStatus.INTERNAL_SERVER_ERROR, new HttpHeaders(),
+                HttpStatus.INTERNAL_SERVER_ERROR.getReasonPhrase(),
+                "Response has already been committed. Unable to redirect.");
+        }
+
+        clearAuthenticationAttributes(request);
+
+        if (!responseUrl.contains("error")) {
```

Comment:
seems like a weird way of checking if the request was successful

Reply:
@dfourn this request will return a 302 with potential error code on it